### PR TITLE
X12Rule to check for duplicate document numbers on an EDI message

### PR DESCRIPTION
**Pull request by CodeSmell**
The X12Rule is designed to be generic and reusable
The current idea is to be able to wire a rule set into the EDI splitter to allow a consumer to verify some basic values in the entire message that would be harder to do after splitting

```
splitter.registerX12Rule(new UniqueDocumentX12Rule(“BSN”, 2, 3));
splitter.registerX12Rule(new UniqueDocumentX12Rule(“BEG”, 3, 4));
splitter.registerX12Rule(new UniqueDocumentX12Rule(“YYZ”, 33 93));
splitter.registerX12Rule(new TrailerSegmentCountX12Rule());
```

These rules would be chained in the splitter and be run one after the other

### CODE REVIEW

File: src/main/java/com/walmartlabs/x12/rule/UniqueDocumentX12Rule.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    public void verify(List<X12Segment> segmentList) {
+        Set<String> documentNumbers = new HashSet<>();
+
+        SegmentIterator segments = new SegmentIterator(segmentList);
+        while (segments.hasNext()) {
+            X12Segment currentSegment = segments.next();
+            if (segmentId.equals(currentSegment.getIdentifier())) {
+                String docNumber = currentSegment.getElement(documentNumberIndex);
+                String docDate = currentSegment.getElement(documentDateIndex);
+                if (!documentNumbers.add(docNumber + "_" + docDate)) {
```
Comment (jaympatel) [inline src/main/java/com/walmartlabs/x12/rule/UniqueDocumentX12Rule.java:50]: Should we do null check for `docNumber` & `docDate`?

Comment (jaympatel) [reply src/main/java/com/walmartlabs/x12/rule/UniqueDocumentX12Rule.java:50]: Should this be checked for transaction set type, document number & document date? 🤔

Comment (patelch) [reply src/main/java/com/walmartlabs/x12/rule/UniqueDocumentX12Rule.java:50]: Are they optional?

**Review by patelch — APPROVED** [review]

File: src/main/java/com/walmartlabs/x12/rule/UniqueDocumentX12Rule.java (hunk 1 above)
Comment (CodeSmell) [reply src/main/java/com/walmartlabs/x12/rule/UniqueDocumentX12Rule.java:50]: will add some tests to demonstrate what happens when these values are missing, but not going to check for null/missing values. 

The reason for that, is that as a general principle: I really don't think that the X12Rule provided by Gozer (at least not those used by the splitter) should be doing much validation and evaluation beyond spot checking the whole EDI message (the main elements are in the correct order (ISA/IEA, GS/GE, and ST/SE) and things that would need to be done across transactions. Most of the other validation should be done on an individual document/transaction after it has been parsed. This allows one transaction to be rejected from the EDI message, while another can be processed. 

For example, I don't think by default the entire message should be rejected if a document number is missing on a single document/transaction. However, if two document/transactions are both missing document numbers that is flagged as an issue (there are two transactions that are missing the value). 

That said, consumers/users of the Gozer framework can easily create their own rules for different behaviors b/c the splitter and the rules were designed with extensibility in mind

**Review by patelch — APPROVED** [review]

**Review by jaympatel — APPROVED** [review]
